### PR TITLE
Implement ReceiveLocalSlamResultsHandler.

### PR DESCRIPTION
**Pull request by cschuet**

### CODE REVIEW

File: cartographer_grpc/proto/map_builder_service.proto [hunk 1]
```diff
@@ -79,6 +93,11 @@ service MapBuilderService {
   rpc AddFixedFramePoseData(stream AddFixedFramePoseDataRequest)
       returns (google.protobuf.Empty);
 
+  // Request a server-side stream of local SLAM results for the given
```
Comment (gaschler) [inline cartographer_grpc/proto/map_builder_service.proto]: "Request the server to send a stream"

File: cartographer/sensor/range_data.cc [hunk 2]
```diff
@@ -54,5 +54,34 @@ TimedRangeData CropTimedRangeData(const TimedRangeData& range_data,
...
 }
 
+proto::RangeData ToProto(const RangeData& range_data) {
+  proto::RangeData proto;
+  *proto.mutable_origin() = transform::ToProto(range_data.origin);
+  proto.mutable_returns()->Reserve(range_data.returns.size());
+  for (const auto& point : range_data.returns) {
+    *proto.add_returns() = transform::ToProto(point);
+  }
+  proto.mutable_misses()->Reserve(range_data.misses.size());
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/range_data.cc]: as discussed this is not going to be more readable.

File: cartographer/sensor/range_data.cc [hunk 3]
```diff
@@ -54,5 +54,34 @@ TimedRangeData CropTimedRangeData(const TimedRangeData& range_data,
...
+  for (const auto& point : range_data.misses) {
+    *proto.add_misses() = transform::ToProto(point);
+  }
+  return proto;
+}
+
+RangeData FromProto(const proto::RangeData& proto) {
+  PointCloud returns;
+  std::transform(
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer/sensor/range_data.cc]: Done.

File: cartographer_grpc/handlers/receive_local_slam_results_handler.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+class ReceiveLocalSlamResultsHandler
+    : public framework::RpcHandler<
+          proto::ReceiveLocalSlamResultsRequest,
+          framework::Stream<proto::ReceiveLocalSlamResultsResponse>> {
+ public:
+  void OnRequest(
+      const proto::ReceiveLocalSlamResultsRequest& request) override {
+    auto writer = GetWriter();
+
```
[the remark this answers is not in the document]
Comment (cschuet) [reply cartographer_grpc/handlers/receive_local_slam_results_handler.h]: Done.

File: cartographer_grpc/proto/map_builder_service.proto (hunk 1 above)
Comment (cschuet) [reply cartographer_grpc/proto/map_builder_service.proto]: Done.

**Comment by cschuet** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user cschuet. Merge queue now has a length of 2.

**Comment by ojura** [comment]
@SirVer @wohe @cshuet We've had `CompressedRangeData` before for efficiently serializing range data. In this PR serializing range data has been reimplemented, but without compression. Bring back `CompressedRangeData`?

**Comment by gaschler** [comment]
Note that `CompressedRangeData` does not store relative time by point.

**Pull request merged**